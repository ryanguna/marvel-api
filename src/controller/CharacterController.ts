/**
 * Module Dependencies
 */

import { Request, Response, NextFunction } from 'express';
import _ from 'lodash';
import Bluebird from 'bluebird';
import { StatusCodes } from 'http-status-codes';

import CharacterService from 'service/CharacterService';
import CharacterTransformer from 'transformer/CharacterTransformer';
import RedisService from 'service/RedisService';
import NotFoundError from 'error/NotFoundError';

export default class CharacterController {
  /**
   * Handler for GET /characters
   * Returns all characters from cache or Marvel API request
   *
   * @static
   * @param request Request
   * @param response Response
   * @param next NextFunction
   * @returns Bluebird<void>
   * @memberOf CharacterController
   */
  public static async index(
    request: Request,
    response: Response,
    next: NextFunction,
  ): Bluebird<void> {
    const url = request.originalUrl;
    const limit = _.get(request.query, 'limit', 100) as number;
    const offset = _.get(request.query, 'offset', 0) as number;

    await Bluebird.try(() => RedisService.getEtag(url))
      .then(etag => CharacterService.getAllCharacters(limit, offset, etag))
      .then(async ({ statusCode, body }) => {
        switch (statusCode) {
          case StatusCodes.OK:
            await RedisService.setEtag(url, body.etag);
            await RedisService.setCachedData(url, body.data.results);
            return body.data.results;
          case StatusCodes.NOT_MODIFIED:
            return RedisService.getCachedData(url);
          default:
            throw new Error(`Unhandled API status code ${body}`);
        }
      })
      .then(characters => CharacterTransformer.transformCharacters(characters))
      .then(payload => response.status(StatusCodes.OK).json(payload))
      .catch(err => next(err));
  }

  /**
   * Handler for GET /characters
   * Returns all characters from cache or Marvel API request
   *
   * @static
   * @async
   * @param request Request
   * @param response Response
   * @param next NextFunction
   * @throws NotFoundError
   * @returns Bluebird<void>
   * @memberOf CharacterController
   */
  public static async show(
    request: Request,
    response: Response,
    next: NextFunction,
  ): Bluebird<void> {
    const characterId = parseInt(request.params.id, 10);

    await Bluebird.try(() => CharacterService.getCharacter(characterId))
      .then(({ statusCode, body }) => {
        switch (statusCode) {
          case StatusCodes.OK:
            return body.data.results;
          case StatusCodes.NOT_FOUND:
            throw new NotFoundError(`Resource id ${characterId} not found`);
          default:
            throw new Error(`Unhandled API status code ${body}`);
        }
      })
      .then(characters => CharacterTransformer.transformCharacter(characters))
      .then(payload => response.status(StatusCodes.OK).json(payload))
      .catch(err => next(err));
  }
}
