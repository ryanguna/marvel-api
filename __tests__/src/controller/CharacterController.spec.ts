import { Request, Response } from 'express';

import CharacterController from 'controller/CharacterController';

const RedisService = require('service/RedisService').default;
const CharacterService = require('service/CharacterService').default;
const CharacterTransformer = require('transformer/CharacterTransformer')
  .default;

jest.mock('service/RedisService');
jest.mock('service/CharacterService');
jest.mock('transformer/CharacterTransformer');

describe('service/CharacterController', () => {
  beforeEach(() => {
    // Clear all instances and calls to constructor and all methods:
    RedisService.getEtag.mockClear();
    RedisService.setEtag.mockClear();
    RedisService.setCachedData.mockClear();
    RedisService.getCachedData.mockClear();
    CharacterService.getAllCharacters.mockClear();
    CharacterService.getCharacter.mockClear();
    CharacterTransformer.transformCharacters.mockClear();
    CharacterTransformer.transformCharacter.mockClear();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
  describe('#index()', () => {
    it('Should save to cache and return data if marvel api status is 200', async () => {
      const request = {
        method: 'GET',
      };

      const response = {};
      const next = jest.fn(() => ({}));

      CharacterService.getAllCharacters.mockImplementationOnce(() => ({
        statusCode: 200,
        body: {
          data: {
            results: [],
          },
        },
      }));

      await CharacterController.index(
        request as Request,
        response as Response,
        next,
      );

      expect(RedisService.getEtag).toHaveBeenCalledTimes(1);
      expect(RedisService.setEtag).toHaveBeenCalledTimes(1);
      expect(RedisService.setCachedData).toHaveBeenCalledTimes(1);
      expect(CharacterTransformer.transformCharacters).toHaveBeenCalledTimes(1);

      // If marvel api returns 200 we dont expect to get data from cache
      expect(RedisService.getCachedData).toHaveBeenCalledTimes(0);
    });

    it('Should read from cache and return data if marvel api status is 304', async () => {
      const request = {
        method: 'GET',
      };

      const response = {};
      const next = jest.fn(() => ({}));

      CharacterService.getAllCharacters.mockImplementationOnce(() => ({
        statusCode: 304,
        body: {
          data: {
            results: [],
          },
        },
      }));

      await CharacterController.index(
        request as Request,
        response as Response,
        next,
      );

      expect(RedisService.getEtag).toHaveBeenCalledTimes(1);
      expect(RedisService.getCachedData).toHaveBeenCalledTimes(1);
      expect(CharacterTransformer.transformCharacters).toHaveBeenCalledTimes(1);

      // If marvel api returns 304 we dont expect to cache data
      expect(RedisService.setEtag).toHaveBeenCalledTimes(0);
      expect(RedisService.setCachedData).toHaveBeenCalledTimes(0);
    });

    it('Should resolve with an error if unhandled marvel api status', async () => {
      const request = {
        method: 'GET',
      };

      const response = {};
      const next = jest.fn(() => ({}));

      CharacterService.getAllCharacters.mockImplementationOnce(() => ({
        statusCode: 400,
        body: {
          data: {
            results: [],
          },
        },
      }));

      await expect(
        CharacterController.index(
          request as Request,
          response as Response,
          next,
        ),
      ).resolves.toMatchSnapshot();
    });
  });
});
