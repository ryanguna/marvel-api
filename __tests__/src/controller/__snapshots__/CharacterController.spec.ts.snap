// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`service/CharacterController #index() Should resolve with an error if unhandled marvel api status 1`] = `undefined`;
